src/sentry/sentry_metrics/querying/api.py

```diff
@@ -6,94 +7,214 @@
-) -> Generator[Timeseries, None, None]:
-    """
-    This function supports parsing in the form:
-    aggregate(metric_name)
-    """
-    if not fields:
-        raise InvalidMetricsQuery("You must query at least one field.")
+@dataclass
+class GroupValue:
+    series: Series
+    total: Total
 
-    for field in fields:
-        match = FIELD_REGEX.match(field)
-        if match is None:
-            raise InvalidMetricsQuery(f"The field {field} can't be parsed.")
+    @classmethod
+    def empty(cls) -> "GroupValue":
+        return GroupValue(series=[], total=None)
 
-        aggregate = match.group(1)
-        metric_name = match.group(2)
+    def add_series_entry(self, time: str, aggregate_value: ResultValue):
+        self.series.append((time, aggregate_value))
+
+    def add_total(self, total: ResultValue):
+        self.total = total
 
-        if is_mri(metric_name):
-            metric = Metric(mri=metric_name)
-        else:
-            metric = Metric(public_name=metric_name)
 
-        yield Timeseries(metric=metric, aggregate=aggregate, filters=filters, groupby=groupby)
+@dataclass(frozen=True)
+class ExecutionResult:
+    # This refers to the timeseries query, not the totals.
+    query: MetricsQuery
+    result: Mapping[str, Any]
+    with_totals: bool
+
+    @property
+    def query_name(self) -> str:
+        timeseries = self.query.query
+        return f"{timeseries.aggregate}({timeseries.metric.mri or timeseries.metric.public_name})"
+
+    @property
+    def modified_start(self) -> datetime:
+        return self.result["modified_start"]
+
+    @property
+    def modified_end(self) -> datetime:
+        return self.result["modified_end"]
+
+    @property
+    def interval(self) -> int:
+        return self.query.rollup.interval
+
+    @property
+    def group_bys(self) -> Sequence[str]:
+        group_bys = []
+        for group_by in self.query.query.groupby or ():
+            if isinstance(group_by, Column):
+                group_bys.append(group_by.name)
+            elif isinstance(group_by, AliasedExpression):
+                group_bys.append(group_by.exp.name)
+
+        return group_bys
+
+    @property
+    def data(self) -> Mapping[str, Any]:
+        return self.result["data"]
+
+    @property
+    def totals(self) -> Mapping[str, Any]:
+        return self.result["totals"]
+
+    @property
+    def meta(self) -> Sequence[Mapping[str, str]]:
+        return self.result["meta"]
+
+
+class QueryParser:
+    def __init__(
+        self,
+        fields: Sequence[str],
+        query: Optional[str],
+        group_bys: Optional[Sequence[str]],
+    ):
+        self._fields = fields
+        self._query = query
+        self._group_bys = group_bys
+
+    def _parse_query(self) -> Optional[Sequence[Condition]]:
+        """
+        This function supports parsing in the form:
+        key:value (_ key:value)?
+        in which the only supported operator is AND until this logic is switched to the metrics layer.
```

Comment:
I think the MQL parser can parse the entire string? Or is there another format beyond MQL that this is parsing?

Reply:
The query here, is passed with discover grammar. Right now it does only support basic key:value pairs.

From my understanding of the EBNF used in MQL, we support filters but they work a bit differently. For this reason right now we have two options:
1. We parse the discover grammar and compute a set of conditions from it. Using/copying how query builders do it.
2. We traspile the discover grammar to MQL, but I have to check if we can have 1:1 mapping of common ops.

If that is the case I would prefer to go for alternative 2, since that would mean that we could avoid any query builder dependency.

Reply:
The reasoning for why query is different is because the front end uses the search component that is also used across the product. This component has its own grammar.

To avoid breaking it, we would like to support this grammar. However, the end goal would be to somehow have the API either accept MQL directly or like I said above, transpile to MQL.

Reply:
I think option 2 is the better one as well. It really simplifies our logic if we can keep one consistent language for metrics.